README.md

# ggcorrplot2

[![Build
Status](https://travis-ci.org/caijun/ggcorrplot2.svg?branch=master)](https://travis-ci.org/caijun/ggcorrplot2)

Implementation of corrplot using ggplot2

## Introduction

Reinventing wheels is not what I like doing.
[corrplot](https://CRAN.R-project.org/package=corrplot) is a great R
package, but I am really tired of customizing the appearance of
corrplot, for example, the space between colorbar and its tick labels,
the space around the plot that I don’t know how to control when writing
it to PDF on my macOS. This is most likely because I am more familiar
with the Grammar of Graphics implemented in ggplot2 than the base
plotting system in R. There are several R packages (e.g.,
[ggcorrplot](https://github.com/kassambara/ggcorrplot) developed by
Alboukadel Kassambara, [ggcorr](https://github.com/briatte/ggcorr)
developed by François Briatte) that can visualize a correlation matrix
into a
[corrgram](https://www.tandfonline.com/doi/abs/10.1198/000313002533)
using ggplot2; however, they are unable to visualize a correlation
matrix using ellipse and mixed methods. **ggcorrplot2** has implemented
only a subset of features of **corrplot** to meet my urgent needs. See
examples in the **Getting started** section. More functionality will be
added in the future as needed.

## Installation

Get the development version from github:

``` r
if (!requireNamespace("devtools")) install.packages("devtools")
devtools::install_github("caijun/ggcorrplot2")
```

## Getting started

The `mtcars` dataset will be used to demonstrate the usages of
**ggcorrplot2**. Most parameters of **ggcorrplot2** functions are the
same as those of **corrplot**. Therefore, it’s easy for users to migrate
from **corrplot** to **ggcorrplot2**.

``` r
library(ggcorrplot2)

data(mtcars)
# Use corr.test() from psych package to calculate the correlation matrix and 
# corresponding p value matrix without adjustment.
library(psych)
ct <- corr.test(mtcars, adjust = "none")
corr <- ct$r
p.mat <- ct$p

# Visualize the correlation matrix
# --------------------------------
# method = "circle" (default)
ggcorrplot(corr)
```

![](figs/README-unnamed-chunk-2-1.png)<!-- -->

``` r
# method = "square"
ggcorrplot(corr, method = "square")
```

![](figs/README-unnamed-chunk-2-2.png)<!-- -->

``` r
# method = "ellipse"
ggcorrplot(corr, method = "ellipse")
```

![](figs/README-unnamed-chunk-2-3.png)<!-- -->

``` r
# method = "number", display the correlation coefficients
ggcorrplot(corr, method = "number")
```

![](figs/README-unnamed-chunk-2-4.png)<!-- -->

``` r
# Visualize the upper or lower triangle of correlation matrix
# -----------------------------------------------------------
# the upper triangle
ggcorrplot(corr, type = "upper")
```

![](figs/README-unnamed-chunk-2-5.png)<!-- -->

``` r
# the lower triangle
ggcorrplot(corr, type = "lower")
```

![](figs/README-unnamed-chunk-2-6.png)<!-- -->

``` r
# Visualize the correlation matrix using mixed methods
# ----------------------------------------------------
# default: upper = "circle", lower = "number"
ggcorrplot.mixed(corr)
```

![](figs/README-unnamed-chunk-2-7.png)<!-- -->

``` r
# upper = "ellipse", lower = "number"
ggcorrplot.mixed(corr, upper = "ellipse", lower = "number")
```

![](figs/README-unnamed-chunk-2-8.png)<!-- -->

``` r
# Combine correlogram with the significance test
# ----------------------------------------------
# Insignificant coefficients according to the default significant level 
# (sig.lvl = 0.05) are indicated by X by default.
ggcorrplot.mixed(corr, upper = "ellipse", lower = "number", p.mat = p.mat)
```

![](figs/README-unnamed-chunk-2-9.png)<!-- -->

``` r
# Leave blank on insignificant coefficients
ggcorrplot.mixed(corr, upper = "ellipse", lower = "number", p.mat = p.mat, 
                 insig = "blank")
```

![](figs/README-unnamed-chunk-2-10.png)<!-- -->

``` r
# Label significant coefficients with asterisks (*, default) denoting the significance level
ggcorrplot.mixed(corr, upper = "ellipse", lower = "number", p.mat = p.mat, 
                 insig = "label_sig", sig.lvl = c(0.05, 0.01, 0.001))
```

![](figs/README-unnamed-chunk-2-11.png)<!-- -->

``` r
# Label significant coefficients with varying number of + denoting the significance level
(p <- ggcorrplot.mixed(corr, upper = "ellipse", lower = "number", p.mat = p.mat, 
                 insig = "label_sig", sig.lvl = c(0.05, 0.01, 0.001), pch = "+", 
                 pch.cex = 4))
```

![](figs/README-unnamed-chunk-2-12.png)<!-- -->

The above examples reproduce some features of **corrplot**. In the
following example, the added advantages of implementing **corrplot**
using ggplot2, such as customizing the appearance of corrgram, combining
a corrgram with other plots (including non-corrgrams) into one plot
using [cowplot](https://github.com/wilkelab/cowplot), are demonstrated.

``` r
# Customize the appearance of corrplot using functions from ggplot2
library(ggplot2)
# Use different color palette
col1 <- colorRampPalette(c("#7F0000", "red", "#FF7F00", "yellow", "white",
                           "cyan", "#007FFF", "blue", "#00007F"))
# Change the colorbar direction to horizontal and place it at the bottom
# As mixed methods are used, there are two scales: color filled in ellipse and 
# number color
p <- p + scale_fill_gradientn(colours = col1(10), limits = c(-1, 1),
                                guide = guide_colorbar(
                                  direction = "horizontal",
                                  title = "",
                                  nbin = 1000,
                                  ticks.colour = "black",
                                  frame.colour = "black",
                                  barwidth = 15,
                                  barheight = 1.5)) +
  scale_colour_gradientn(colours = col1(10), limits = c(-1, 1),
                       guide = guide_colorbar(
                         direction = "horizontal",
                         title = "",
                         nbin = 1000,
                         ticks.colour = "black",
                         frame.colour = "black",
                         barwidth = 15,
                         barheight = 1.5)) +
  theme(legend.position = "bottom")
p
```

<img src="figs/README-unnamed-chunk-3-1.png" style="display: block; margin: auto;" />

``` r
# Combine a lower corrgram and a mixed corrgram side by side with a shared colorbar on the bottom

# NAs are allowed in correlation matrix and p value matrix, which are labelled as NA
# Assign NAs
rid <- c(2, 3, 1, 2)
cid <- c(3, 2, 2, 1)
pos <- cbind(rid, cid)
corr[pos] <- NA
p.mat[pos] <- NA

# a lower corrgram
p1 <- ggcorrplot(corr, type = "lower", method = "square", p.mat = p.mat, 
                 insig = "label_sig", sig.lvl = c(0.05, 0.01, 0.001), show.diag = FALSE)
# a mixed corrgram
p2 <- ggcorrplot.mixed(corr, upper = "ellipse", lower = "number", p.mat = p.mat, 
                       insig = "label_sig", sig.lvl = c(0.05, 0.01, 0.001), 
                       pch = "+", pch.cex = 4)

library(cowplot)
prow <- plot_grid(p1 + theme(legend.position = "none"),
                  p2 + theme(legend.position = "none"),
                  rel_widths = c(1, 1), nrow = 1, align = 'hv',
                  labels = c("(a)", "(b)"), label_x = 0, label_y = 1)

# Extract the legend from the first corrgram
legend <- get_legend(p1)
# Add the legend to the bottom of the plot row we made earlier.
p <- cowplot::plot_grid(prow, legend, ncol = 1, rel_heights  = c(1, 0.15))
p
```

<img src="figs/README-unnamed-chunk-3-2.png" style="display: block; margin: auto;" />

## Citation

To cite the ‘ggcorrplot2’ package in publications use:

``` 
  Jun Cai (2019). ggcorrplot2: Visualize a Correlation Matrix using ggplot2. R package version 0.1.0.
```

A BibTeX entry for LaTeX users is

``` 
  @Manual{,
    title = {ggcorrplot2: Visualize a Correlation Matrix using ggplot2},
    author = {Jun Cai},
    year = {2019},
    note = {R package version 0.1.0},
    url = {https://github.com/caijun/ggcorrplot2},
  }
```

The above citation information can be generated by calling
`citation("ggcorrplot2")` in R.

## Contact

Bugs and feature requests can be filed to
<https://github.com/caijun/ggcorrplot2/issues>. Pull requests are also
welcome.


NEWS.md
### CHANGES IN ggcorrplot VERSION 0.1.0

#### NEW FEATURES

- first version of ggcorrplot

#### NOTE

- in this NEWS file, #n means the issue number on GitHub, e.g. #1 is https://github.com/caijun/ggcorrplot/issues/1


test.R
rm(list = ls())

source("ggcorrplot.mixed.R")

# test
corr <- cor(mtcars)
p.mat <- cor.mtest(mtcars, conf.level = 0.95)

p <- ggcorrplot(corr = corr, p.mat = p.mat, method = "ellipse", type = "full", 
                show.diag = TRUE)
p

p <- ggcorrplot.mixed(corr = corr, p.mat = p.mat, upper = "ellipse", 
                      lower = "number")
p

R/ggcorrplot.mixed.R
#' @title Visualize a correlation matrix using mixed methods
#'
#' @param corr a correlation matrix to be visualized
#' @param upper a character indicating the visualization method of the upper triangular matrix to be used. Currently, it supports four methods, named \code{"circle"} (default), \code{"square"}, \code{"ellipse"}, \code{"number"}.
#' @param lower a character indicating the visualization method of the lower triangular matrix to be used. Currently, it supports four methods, named \code{"circle"}, \code{"square"}, \code{"ellipse"}, \code{"number"}(default).
#' @param p.mat a matrix of p-value
#' @param sig.lvl a numeric vector specifying significant level(s). If the p-value in \code{p.mat} is bigger than \code{sig.lvl} (0.05 by default), then the corresponding correlation coefficient is regarded as insignificant. If \code{insig} is \code{"label_sig"}, this may be an increasing vector of significance levels, for example \code{c(0.05, 0.01, 0.001)}, in which case \code{pch} will be used once for the highest p-value interval and multiple times (e.g. "*", "**", "***") for each lower p-value interval.
#' @param number.digits the number of decimal digits (2 by default) while the visualization method is \code{"number"}.
#' @param insig a character specialized insignificant correlation coefficients, \code{"pch"} (default), \code{"blank"}, or \code{"label_sig"}.
#' @param pch a point character indicating the shape of insignificant correlation coefficients.
#' @param pch.cex a number controlling the shape size of insignificant correlation coefficients.
#' @export
ggcorrplot.mixed <- function(corr, upper = c("circle", "square", "ellipse", "number"),
                             lower = c("number", "square", "ellipse", "circle"),
                             p.mat = NULL, sig.lvl = 0.05, number.digits = 2,
                             insig = c("pch", "blank", "label_sig"),
                             pch = 4, pch.cex = 5) {
  upper <- match.arg(upper)
  lower <- match.arg(lower)
  insig <- match.arg(insig)

  if (is.null(colnames(corr))) {
    stop("Correlation matrix needs column names!")
  }
  vars <- colnames(corr)
  # number of variables
  nvars <- length(vars)

  corr <- reshape2::melt(corr, value.name = "rho") %>%
    mutate(rid = as.integer(as.factor(.data$Var1)),
           cid = as.integer(as.factor(.data$Var2))) %>%
    mutate(part = case_when(
      .$rid < .$cid ~ "upper",
      .$rid == .$cid ~ "diag",
      .$rid > .$cid ~ "lower"
    )) %>%
    mutate(abs.rho = abs(.data$rho)) %>%
    mutate(rho.label = ifelse(.data$rho == 1, .data$rho, format(round(.data$rho, digits = number.digits),
                                                                nsmall = number.digits)))

  if (!is.null(p.mat)) {
    p.mat <- reshape2::melt(p.mat, value.name = "pval") %>%
      mutate(rid = as.integer(as.factor(.data$Var1)),
             cid = as.integer(as.factor(.data$Var2))) %>%
      mutate(part = case_when(
        .$rid < .$cid ~ "upper",
        .$rid == .$cid ~ "diag",
        .$rid > .$cid ~ "lower"
      ))

    sig.codes <- sapply(seq_along(sig.lvl), function(i) {
      # By default, mark significance with *
      if (!is.character(pch)) {
        pch <- "*"
      }
      paste(rep(pch, i), collapse = "")
    })

    corr <- corr %>%
      mutate(pval = p.mat$pval) %>%
      mutate(signif = as.numeric(.data$pval <= max(sig.lvl))) %>%
      mutate(sig.codes = cut(.data$pval, breaks = c(sig.lvl, 0, 1), labels = c(rev(sig.codes), ""),
                             include.lowest = TRUE))

    # insignificant p value matrix
    p.mat.insig <- p.mat %>%
      dplyr::filter(.data$pval > max(sig.lvl))

    if (insig == "blank") {
      corr <- corr %>%
        mutate(rho = .data$rho * signif)
    }
  } else {
    p.mat.insig <- NULL
  }

  # default palette of corrplot
  col2 <- grDevices::colorRampPalette(RColorBrewer::brewer.pal(n = 11, name = "RdBu"))

  p <- ggplot(data = corr) +
    geom_rect(mapping = aes(xmin = .data$cid - 0.5, xmax = .data$cid + 0.5,
                            ymin = .data$rid - 0.5, ymax = .data$rid + 0.5),
              color = "grey92", fill = NA) +
    coord_fixed() +
    scale_x_continuous(expand = c(0, 0)) +
    scale_y_reverse(expand = c(0, 0)) +
    theme_bw() +
    theme(legend.margin = margin(0, unit = 'cm'),
          axis.text.x = element_blank(),
          axis.text.y = element_blank(),
          axis.title = element_blank(),
          axis.ticks = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank())

  if (any(is.na(corr$rho))) {
    corr.NA <- corr %>%
      dplyr::filter(is.na(.data$rho))
    corr <- corr %>%
      dplyr::filter(!is.na(.data$rho))
  } else {
    corr.NA <- NULL
  }

  # add upper plot
  upper.dat <- corr %>%
    dplyr::filter(.data$part == "upper")

  p <- plot.method(p, data = upper.dat, method = upper)
  # add significant codes except number method
  if (!is.null(p.mat) & insig == "label_sig" & upper != "number") {
    p <- p +
      geom_text(data = upper.dat, mapping = aes(x = .data$cid, y = .data$rid), label = upper.dat$sig.codes,
                size = pch.cex)
  }

  # add lower plot
  lower.dat <- corr %>%
    dplyr::filter(.data$part == "lower")

  p <- plot.method(p, data = lower.dat, method = lower)
  # add significant codes except number method
  if (!is.null(p.mat) & insig == "label_sig" & lower != "number") {
    p <- p +
      geom_text(data = lower.dat, mapping = aes(x = .data$cid, y = .data$rid), label = upper.dat$sig.codes,
                size = pch.cex)
  }

  # indicate insigificant p value with point character
  if (!is.null(p.mat.insig) & insig == "pch") {
    p <- p + geom_point(data = p.mat.insig, mapping = aes(x = .data$cid, y = .data$rid),
                        shape = pch, size = pch.cex)
  }

  # indicate NA correlation coefficient
  if (!is.null(corr.NA)) {
    p <- p +
      geom_text(data = corr.NA, mapping = aes(x = .data$cid, y = .data$rid), label = "NA",
                size = pch.cex, color = "grey92")
  }

  # colorbar
  p <- p + scale_fill_gradientn(colours = col2(200), limits = c(-1, 1),
                                guide = guide_colorbar(
                                  title = "",
                                  nbin = 1000,
                                  ticks.colour = "black",
                                  frame.colour = "black",
                                  barwidth = 1.5,
                                  barheight = 15)) +
    scale_colour_gradientn(colours = col2(200), limits = c(-1, 1),
                           guide = guide_colorbar(
                             title = "",
                             nbin = 1000,
                             ticks.colour = "black",
                             frame.colour = "black",
                             barwidth = 1.5,
                             barheight = 15))

  diag.vars <- data.frame(x = 1:nvars, y = 1:nvars)
  diag.label <- vars
  p <- p +
    geom_text(data = diag.vars, mapping = aes(.data$x, .data$y), label = diag.label, colour = "grey30")

  return(p)
}


R/ggcorrplot.R
# internal
# @title Add the geom of correlation matrix to the existing plot
#
# @param p a ggplot object
# @param data a data.frame to be plotted
# @param method a character indicating the visualization method of correlation matrix to be used. Currently, it supports four methods, named "circle" (default), "square", "ellipse", "number".
#' @import ggplot2
#' @importFrom dplyr %>% mutate group_by group_modify case_when
#' @importFrom rlang .data
plot.method <- function(p, data, method = c("circle", "square", "ellipse", "number")) {
  if (method == "ellipse") {
    ellipse.xy <- function(rho, length = 100) {
      theta <- seq(0, 2 * pi, length = length)
      if (rho == 1) rho <- rho - 1e-4
      d <- acos(rho)
      x <- cos(theta + d / 2) / 2
      y <- cos(theta - d / 2) / 2
      as.data.frame(cbind(x, y))
    }

    myfun <- function(df) {
      res <- ellipse.xy(df$rho) %>%
        mutate(rid = df$rid, cid = df$cid, rho =  df$rho) %>%
        # using .data prevents R CMD check from giving a NOTE about undefined global variable
        mutate(x1 = 0.9 * .data$x + .data$cid, y1 = 0.9 * .data$y + .data$rid,
               group = paste(.data$rid, .data$cid, sep = "-"))
    }

    ellipse.dat <- data %>%
      group_by(.data$Var1, .data$Var2) %>%
      group_modify(~myfun(.x))

    p <- p +
      geom_polygon(data = ellipse.dat,
                   mapping = aes(x = .data$x1, y = .data$y1, fill = .data$rho,
                                 group = .data$group),
                   color = NA)
  } else if (method == "circle") {
    p <- p +
      ggforce::geom_circle(data = data,
                           mapping = aes(x0 = .data$cid, y0 = .data$rid,
                                         r = .data$abs.rho/2 - 0.02, fill = .data$rho),
                           color = NA)
  } else if (method == "square") {
    p <- p +
      geom_rect(data = data, mapping = aes(xmin = .data$cid - 0.5*(.data$abs.rho - 0.04),
                                           xmax = .data$cid + 0.5*(.data$abs.rho - 0.04),
                                           ymin = .data$rid - 0.5*(.data$abs.rho - 0.04),
                                           ymax = .data$rid + 0.5*(.data$abs.rho - 0.04),
                                           fill = .data$rho))
  } else if (method == "number") {
    p <- p +
      geom_text(data = data, mapping = aes(x = .data$cid, y = .data$rid, colour = .data$rho),
                label = data$rho.label)
  }
  return(p)
}

#' @title Visualize a correlation matrix
#'
#' @param corr a correlation matrix to be visualized
#' @param method a character indicating the visualization method of correlation matrix to be used. Currently, it supports four methods, named \code{"circle"} (default), \code{"square"}, \code{"ellipse"}, \code{"number"}.
#' @param type a character indicating that the \code{"full"} (default), \code{"upper"} or \code{"lower"} triangular matrix is displayed.
#' @param p.mat a matrix of p-value
#' @param sig.lvl a numeric vector specifying significant level(s). If the p-value in \code{p.mat} is bigger than \code{sig.lvl} (0.05 by default), then the corresponding correlation coefficient is regarded as insignificant. If \code{insig} is \code{"label_sig"}, this may be an increasing vector of significance levels, for example \code{c(0.05, 0.01, 0.001)}, in which case \code{pch} will be used once for the highest p-value interval and multiple times (e.g. "*", "**", "***") for each lower p-value interval.
#' @param number.digits the number of decimal digits (2 by default) while the visualization method is \code{"number"}.
#' @param show.diag a logical indicating whether display the correlation coefficients on the principal diagonal.
#' @param insig a character specialized insignificant correlation coefficients, \code{"pch"} (default), \code{"blank"}, or \code{"label_sig"}.
#' @param pch a point character indicating the shape of insignificant correlation coefficients.
#' @param pch.cex a number controlling the shape size of insignificant correlation coefficients.
#' @export
ggcorrplot <- function(corr, method = c("circle", "square", "ellipse", "number"),
                       type = c("full", "lower", "upper"), p.mat = NULL,
                       sig.lvl = 0.05, number.digits = 2,
                       show.diag = TRUE, insig = c("pch", "blank", "label_sig"),
                       pch = 4, pch.cex = 5) {
  method <- match.arg(method)
  type <- match.arg(type)
  insig <- match.arg(insig)

  if (is.null(colnames(corr))) {
    stop("Correlation matrix needs column names!")
  }
  vars <- colnames(corr)
  # number of variables
  nvars <- length(vars)

  corr <- reshape2::melt(corr, value.name = "rho") %>%
    mutate(rid = as.integer(as.factor(.data$Var1)),
           cid = as.integer(as.factor(.data$Var2))) %>%
    mutate(part = case_when(
      .$rid < .$cid ~ "upper",
      .$rid == .$cid ~ "diag",
      .$rid > .$cid ~ "lower"
    )) %>%
    mutate(abs.rho = abs(.data$rho)) %>%
    mutate(rho.label = ifelse(.data$rho == 1, .data$rho, format(round(.data$rho, digits = number.digits),
                                                                nsmall = number.digits)))

  if (!is.null(p.mat)) {
    p.mat <- reshape2::melt(p.mat, value.name = "pval") %>%
      mutate(rid = as.integer(as.factor(.data$Var1)),
             cid = as.integer(as.factor(.data$Var2))) %>%
      mutate(part = case_when(
        .$rid < .$cid ~ "upper",
        .$rid == .$cid ~ "diag",
        .$rid > .$cid ~ "lower"
      ))

    sig.codes <- sapply(seq_along(sig.lvl), function(i) {
      # By default, mark significance with *
      if (!is.character(pch)) {
        pch <- "*"
      }
      paste(rep(pch, i), collapse = "")
    })

    corr <- corr %>%
      mutate(pval = p.mat$pval) %>%
      mutate(signif = as.numeric(.data$pval <= max(sig.lvl))) %>%
      mutate(sig.codes = cut(.data$pval, breaks = c(sig.lvl, 0, 1), labels = c(rev(sig.codes), ""),
                             include.lowest = TRUE))

    # insignificant p value matrix
    p.mat.insig <- p.mat %>%
      dplyr::filter(.data$pval > max(sig.lvl))

    if (insig == "blank") {
      corr <- corr %>%
        mutate(rho = .data$rho * signif)
    }
  } else {
    p.mat.insig <- NULL
  }

  if (type == "lower") {
    corr <- corr %>%
      dplyr::filter(.data$part != "upper")
    if (!is.null(p.mat.insig)) {
      p.mat.insig <- p.mat.insig %>%
        dplyr::filter(.data$part != "upper")
    }
  } else if(type == "upper") {
    corr <- corr %>%
      dplyr::filter(.data$part != "lower")
    if (!is.null(p.mat.insig)) {
      p.mat.insig <- p.mat.insig %>%
        dplyr::filter(.data$part != "lower")
    }
  }

  if (!show.diag) {
    corr <- corr %>%
      dplyr::filter(.data$part != "diag")
  }

  # default palette of corrplot
  col2 <- grDevices::colorRampPalette(RColorBrewer::brewer.pal(n = 11, name = "RdBu"))

  p <- ggplot(data = corr) +
    geom_rect(mapping = aes(xmin = .data$cid - 0.5, xmax = .data$cid + 0.5,
                            ymin = .data$rid - 0.5, ymax = .data$rid + 0.5),
              color = "grey92", fill = NA) +
    coord_fixed() +
    theme_bw() +
    theme(legend.margin = margin(0, unit = 'cm'),
          axis.title = element_blank(),
          axis.ticks = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank())

  if (!show.diag & type == "full") {
    p <- p +
      geom_rect(mapping = aes(xmin = 1 - 0.5, xmax = 1 + 0.5,
                              ymin = 1 - 0.5, ymax = 1 + 0.5),
                color = "grey92", fill = NA) +
      geom_rect(mapping = aes(xmin = nvars - 0.5, xmax = nvars + 0.5,
                              ymin = nvars - 0.5, ymax = nvars + 0.5),
                color = "grey92", fill = NA)
  }

  if (any(is.na(corr$rho))) {
    corr.NA <- corr %>%
      dplyr::filter(is.na(.data$rho))
    corr <- corr %>%
      dplyr::filter(!is.na(.data$rho))
  } else {
    corr.NA <- NULL
  }

  p <- plot.method(p, data = corr, method = method)
  # add significant codes except number method
  if (!is.null(p.mat) & insig == "label_sig" & method != "number") {
    p <- p +
      geom_text(data = corr, mapping = aes(x = .data$cid, y = .data$rid), label = corr$sig.codes,
                size = pch.cex)
  }

  # indicate insigificant p value with point character
  if (!is.null(p.mat.insig) & insig == "pch") {
    p <- p + geom_point(data = p.mat.insig, mapping = aes(x = .data$cid, y = .data$rid),
                        shape = pch, size = pch.cex)
  }

  # indicate NA correlation coefficient
  if (!is.null(corr.NA)) {
    p <- p +
      geom_text(data = corr.NA, mapping = aes(x = .data$cid, y = .data$rid), label = "NA",
                size = pch.cex, color = "grey92")
  }

  # colorbar
  if (type %in% c("full", "upper")) {
    if (method %in% c("ellipse", "circle", "square")) {
      p <- p + scale_fill_gradientn(colours = col2(200), limits = c(-1, 1),
                                    guide = guide_colorbar(
                                      title = "",
                                      nbin = 1000,
                                      ticks.colour = "black",
                                      frame.colour = "black",
                                      barwidth = 1.5,
                                      barheight = 15))
    } else {
      p <- p + scale_colour_gradientn(colours = col2(200), limits = c(-1, 1),
                                      guide = guide_colorbar(
                                        title = "",
                                        nbin = 1000,
                                        ticks.colour = "black",
                                        frame.colour = "black",
                                        barwidth = 1.5,
                                        barheight = 15))
    }
  } else if (type %in% c("lower")) {
    if (method %in% c("ellipse", "circle", "square")) {
      p <- p + scale_fill_gradientn(colours = col2(200), limits = c(-1, 1),
                                    guide = guide_colorbar(
                                      direction = "horizontal",
                                      title = "",
                                      nbin = 1000,
                                      ticks.colour = "black",
                                      frame.colour = "black",
                                      barwidth = 15,
                                      barheight = 1.5))
    } else {
      p <- p + scale_colour_gradientn(colours = col2(200), limits = c(-1, 1),
                                      guide = guide_colorbar(
                                        direction = "horizontal",
                                        title = "",
                                        nbin = 1000,
                                        ticks.colour = "black",
                                        frame.colour = "black",
                                        barwidth = 15,
                                        barheight = 1.5))
    }
    p <- p + theme(legend.position = "bottom")
  }

  # variable labels
  axis.text.fontsize <- 11 # in pt
  geom.text.fontsize <- axis.text.fontsize / ggplot2::.pt # in mm
  if (type == "full") {
    p <- p +
      scale_x_continuous(breaks = 1:nvars, labels = vars, expand = c(0, 0)) +
      scale_y_reverse(breaks = 1:nvars, labels = vars, expand = c(0, 0))
  } else if (type == "lower") {
    if (show.diag) {
      y.vars <- data.frame(x = 0, y = 1:nvars)
      diag.vars <- data.frame(x = 1:nvars, y = 0:(nvars - 1))
      y.label <- diag.label <- vars
    } else {
      y.vars <- data.frame(x = 0, y = 2:nvars)
      diag.vars <- data.frame(x = 1:(nvars - 1), y = 1:(nvars - 1))
      y.label <- utils::tail(vars, -1)
      diag.label <- utils::head(vars, -1)
    }
    p <- p +
      geom_text(data = diag.vars, mapping = aes(.data$x, .data$y), label = diag.label,
                colour = "grey30", size = geom.text.fontsize) +
      scale_x_continuous(expand = c(0, 0)) +
      scale_y_reverse(breaks = y.vars$y, labels = y.label, expand = c(0, 0),
                      limits = c(nvars + 0.5, min(diag.vars$y) - 0.5)) +
      theme(axis.text.x = element_blank(),
            axis.text.y = element_text(size = axis.text.fontsize))
  } else if (type == "upper") {
    if (show.diag) {
      x.vars <- data.frame(x = 1:nvars, y = 0)
      diag.vars <- data.frame(x = 0:(nvars - 1), y = 1:nvars)
      x.label <- diag.label <- vars
    } else {
      x.vars <- data.frame(x = 2:nvars, y = 0)
      diag.vars <- data.frame(x = 1:(nvars - 1), y = 1:(nvars - 1))
      x.label <- utils::tail(vars, -1)
      diag.label <- utils::head(vars, -1)
    }
    p <- p +
      geom_text(data = diag.vars, mapping = aes(.data$x, .data$y), label = diag.label,
                colour = "grey30", size = geom.text.fontsize) +
      scale_x_continuous(breaks = x.vars$x, labels = x.label, expand = c(0, 0),
                         position = "top", limits = c(min(diag.vars$x) - 0.5, nvars + 0.5)) +
      scale_y_reverse(expand = c(0, 0)) +
      theme(axis.text.x = element_text(size = axis.text.fontsize),
            axis.text.y = element_blank())
  }

  return(p)
}
